developer-extension/src/background/devtoolsPanelConnection.ts

```diff
@@ -0,0 +1,59 @@
+import type { BackgroundToDevtoolsMessage, DevtoolsToBackgroundMessage } from '../common/types'
+import { EventListeners } from '../common/eventListeners'
+import { createLogger } from '../common/logger'
+
+const logger = createLogger('devtoolsPanelConnection')
+
+const devtoolsConnections = new Map<number, chrome.runtime.Port>()
+
+const portNameRe = /^devtools-panel-for-tab-(\d+)$/
+
+export const onDevtoolsFirstConnection = new EventListeners<number>()
+export const onDevtoolsLastDisconnection = new EventListeners<number>()
```

Comment:
<!-- Thoughts represent an idea that popped up from reviewing. These comments are non-blocking by nature, but they are extremely valuable and can lead to more focused initiatives and mentoring opportunities.  -->
**💭 thought:** ‏a `type TabId = number` alias could make those types more expressives